autoload/suda.vim

```diff
@@ -93,6 +93,33 @@ function! suda#read(expr, ...) abort range
   endtry
 endfunction
 
+function! suda#write_wrapper(write_cmd) abort
+  try 
+    execute(a:write_cmd)
+  catch /^Vim\%((\a\+)\)\=:E45/
+    execute(a:write_cmd . ' suda://%')
+  endtry
+endfunction
+
+let s:smart_read_checked = {}
+
+function! suda#smart_read(bufn, fpath) abort
+  if get(s:smart_read_checked, a:bufn, 0)
+    let s:smart_read_checked[a:bufn] = 1
+    return
+  end
+  if !match(a:fpath, '^[a-z]\+://*') || &buftype != '' || a:fpath ==# ''
+    return
+  endif
+  if !filereadable(a:fpath)
+    if filewritable(fnamemodify(a:fpath, ':p:h')) ==# 2
+      return
+    endif
+    execute 'edit suda://%'
+    execute "bdelete! " . a:bufn
+  endif
+endfunction
+
```

Comment:
**must**

This function does

1. Skip if the buffer has already checked
2. Skip if the buffer name starts from scheme (e.g. `fugitive://`)
3. Skip if the buffer is not a file
4. Skip if the buffer is new one
5. Skip if the file of the buffer does not exist but writable
6. Use `suda://` prefix instead if the buffer is not readable
7. Skip if the buffer is readable

But using `BufReadPre` will solves 1-5 I guess while

1. `BufReadPre` is not used if the file doesn' exist
2. `BufReadPre` is *usually* not used with `BufReadCmd` which is used for scheme loading (e.g. `fugitive://`)

Additionaly, I think using `execute 'keepjumps keepalt file suda://%'` is enough instead for executing `edit/bdelete!` sequentially (not confirmed).

Anyway, using `BufReadPre` would change this function completely I guess so I'll review new one.

Reply:
You have no idea how hard it was for me to find the right event.. I wished `BufReadPre` could be used but it's not and I had my head wrapped around it for a whole day why until I've found this stack-exchange QA that gives a good enough explanation / workaround: https://vi.stackexchange.com/questions/13178/autocmd-bufadd-bufcreate-bufnew-arent-triggered-for-new-unnamed-buffers-ene

Reply:
Now I tried and I got what you meant 👍 

# tl;dr

1. Remove unnecessary autocmd instead of caching the detection result per buffers
2. Use `is#` or `isnot#` to compare numbers
3. Check if the file is not readable or does not exist by `getftype()`
4. Use `keepalt keepjumps file` to replace the buffer name instead

In short, apply the following diff.

```diff
diff --git a/autoload/suda.vim b/autoload/suda.vim
index 8487e31..e444a36 100644
--- a/autoload/suda.vim
+++ b/autoload/suda.vim
@@ -94,32 +94,13 @@ function! suda#read(expr, ...) abort range
 endfunction
 
 function! suda#write_wrapper(write_cmd) abort
-  try 
+  try
     execute(a:write_cmd)
   catch /^Vim\%((\a\+)\)\=:E45:/
     execute(a:write_cmd . ' suda://%')
   endtry
 endfunction
 
-let s:smart_read_checked = {}
-
-function! suda#smart_read(bufn, fpath) abort
-  if get(s:smart_read_checked, a:bufn, 0)
-    let s:smart_read_checked[a:bufn] = 1
-    return
-  end
-  if !match(a:fpath, '^[a-z]\+://*') || &buftype != '' || a:fpath ==# ''
-    return
-  endif
-  if !filereadable(a:fpath)
-    if filewritable(fnamemodify(a:fpath, ':p:h')) ==# 2
-      return
-    endif
-    execute 'edit suda://%'
-    execute "bdelete! " . a:bufn
-  endif
-endfunction
-
 function! suda#write(expr, ...) abort range
   let path = s:expand_expression(a:expr)
   let options = extend({
@@ -161,7 +142,6 @@ function! suda#write(expr, ...) abort range
   endtry
 endfunction
 
-
 function! suda#BufReadCmd() abort
   call s:doautocmd('BufReadPre')
   let ul = &undolevels
@@ -226,6 +206,25 @@ function! suda#FileWriteCmd() abort
   endtry
 endfunction
 
+function! suda#BufEnter() abort
+  augroup suda_smart_read_local
+    autocmd! * <buffer=abuf>
+  augroup END
+  let bufname = expand('<afile>')
+  if !empty(&buftype) || empty(bufname) || match(bufname, '^[a-z]\+://*') isnot# -1
+    " Non file buffer
+    return
+  elseif filereadable(bufname)
+    " File is readable
+    return
+  elseif empty(getftype(bufname)) && filewritable(fnamemodify(bufname, ':p:h')) is# 2
+    " File does not exist and the directory is writable
+    return
+  endif
+  execute printf('keepalt keepjumps file %s%%', g:suda#prefix)
+  edit
+endfunction
+
 function! s:escape_patterns(expr) abort
   return escape(a:expr, '^$~.*[]\')
 endfunction
diff --git a/plugin/suda.vim b/plugin/suda.vim
index 71aec8f..2276acd 100644
--- a/plugin/suda.vim
+++ b/plugin/suda.vim
@@ -15,8 +15,20 @@ if get(g:, 'suda_hijack_write')
 endif
 
 if get(g:, 'suda_smart_read')
-  augroup suda_smart_read
+  augroup suda_smart_read_primary
     autocmd!
-    autocmd BufEnter * nested call suda#smart_read(expand('<abuf>'), expand('<afile>'))
-  augroup end
+    autocmd BufEnter * nested call suda#BufEnter()
+    autocmd VimEnter *
+          \ augroup suda_smart_read_primary |
+          \   autocmd! * |
+          \ augroup END
+  augroup END
+
+  augroup suda_smart_read_secondary
+    autocmd! * <buffer>
+    autocmd BufNew *
+         \ augroup suda_smart_read_local |
+         \   autocmd BufEnter <buffer=abuf> nested call suda#BufEnter() |
+         \ augroup END
+  augroup END
 endif
```

**NOTE** that I haven't try above with all situations you might consider so please test by yourself :+1:

# Experiment

## Environment

I made readable and unreadable files by

```
$ echo "READABLE" > readable
$ echo "UNREADABLE" > unreadable
$ sudo chown root readable unreadable
$ sudo chmod 600 unreadable
```

Then add the following code to my `.vimrc` to investigate events

```vim
augroup autocmd_investigation
  autocmd!
  autocmd BufNew * echomsg "BufNew"
  autocmd BufAdd * echomsg "BufAdd"
  autocmd BufEnter * echomsg "BufEnter"

  autocmd BufNewFile * echomsg "BufNewFile"

  autocmd BufFilePre * echomsg "BufFilePre"
  autocmd BufFilePost * echomsg "BufFilePost"

  autocmd BufReadPre * echomsg "BufReadPre"
  autocmd BufReadPost * echomsg "BufReadPost"

  autocmd FileReadPre * echomsg "FileReadPre"
  autocmd FileReadPost * echomsg "FileReadPost"

  autocmd VimEnter * echomsg "VimEnter"
  autocmd WinEnter * echomsg "WinEnter"
augroup end
```

# Result

## Read readable/unreadable on startup

### Vim
```
$ vim readable
BufReadPre
BufReadPost
BufEnter
VimEnter

$ vim unreadable
BufEnter
VimEnter
```

### Neovim
```
$ nvim readable
BufEnter
VimEnter

$ nvim unreadable
BufEnter
VimEnter
```

## Read readable/unreadable with `:e` command

### Vim

```vim
:e readable
BufNew
BufAdd
BufReadPre
BufReadPost
BufEnter

:e unreadable
BufNew
BufAdd
BufEnter
```

### Neovim

```vim
:e readable
BufNew
BufAdd
BufEnter

:e unreadable
BufNew
BufAdd
BufEnter
```

# Conclusion

It seems there are two distinct pathway. So my conclusion is

1. Use `BufEnter` to detect if argument files are readable or not before `VimEnter` event
2. Use `BufEnter` which is sequentially after `BufNew` event

Reply:
Wow, well done on this investigation. It seems to be functioning well with my initial testings. I think it'll be good to keep this PR open for maybe even a few weeks if you don't mind so I can test it thoroughly with my normal usage with Vim and push small updates as I go.

Now I'd like please just to make sure I understand what you've done in `plugin/suda.vim`:

Are you running on `BufEnter` `suda#BufEnter()` and then remove this `autocmd` (addressing it through the `augroup` named `suda_smart_read_primary`)? If my understanding is correct, that's f*cking creative! Way better then those cheap workarounds in the [stack-exchange thread I mentioned](https://vi.stackexchange.com/questions/13178/autocmd-bufadd-bufcreate-bufnew-arent-triggered-for-new-unnamed-buffers-ene).

If my 1st assumption is correct, then I'd assume you've done a similar thing in `augroup suda_smart_read_secondary` but for `:e` commands and therfor, without `VimEnter` which is not fired. However, I don't understand: Why did you create another `augroup suda_smart_read_local` inside it? Inspired by what you've done in `augroup suda_smart_read_primary` I would do the same for when `:e` is used but using these events:

- Instead of `BufEnter`: `BufNew`
- Instead of `VimEnter`: `BufAdd`

Reply:
>  I think it'll be good to keep this PR open for maybe even a few weeks if you don't mind so I can test it thoroughly with my normal usage with Vim and push small updates as I go.

Totally agree with your suggestion 👍 There might be some edge cases.

> Are you running on BufEnter suda#BufEnter() and then remove this autocmd (addressing it through the augroup named suda_smart_read_primary)? If my understanding is correct, that's f*cking creative! Way better then those cheap workarounds in the stack-exchange thread I mentioned.

Yes. This is a kind hacky way but quite effective. Note that Vim 8.1.1113 or Neovim 0.4.0-dev has `autocmd-once` so this kind of hack will not be required in the future.

> If my 1st assumption is correct, then I'd assume you've done a similar thing in augroup suda_smart_read_secondary but for :e commands and therfor, without VimEnter which is not fired. However, I don't understand: Why did you create another augroup suda_smart_read_local inside it? Inspired by what you've done in augroup suda_smart_read_primary I would do the same for when :e is used but using these events:

Ah, it seems your sense is incredible. Now I think this *once-like hack* is not required for the second one. Using `BufAdd` without using `BufNew` is enough I guess. Anyway, the above code is not effective so please optimize it 👍

Reply:
> Ah, it seems your sense is incredible. Now I think this once-like hack is not required for the second one. Using `BufAdd` without using `BufNew` is enough I guess. Anyway, the above code is not effective so please optimize it :+1:

Thanks :smile:. I must say though, that this hack is so slippery! I wonder how it took so long to feel the need for `autocmd-once`.. Anyway, I tried to rewrite what you did in somewhat more symmetric way for both `:e` and `$EDITOR file` situations:

```vim
if get(g:, 'suda_smart_read')
  augroup suda_smart_read_primary
    autocmd!
    autocmd BufEnter * nested call suda#BufEnter(expand('<afile>'))
    autocmd VimEnter *
          \ augroup suda_smart_read_primary |
          \   autocmd! * |
          \ augroup END
  augroup END

  augroup suda_smart_read_secondary
    autocmd!
    autocmd BufNew * nested call suda#BufEnter(expand('<afile>'))
    autocmd BufAdd *
          \ augroup suda_smart_read_secondary |
          \   autocmd! * |
          \ augroup END
  augroup END
endif
```

But, the problem is, that since we have to use `nested`, and `suda#BufEnter()` calls `edit`, I always get double `suda://` prefixes: `suda://suda://%`. Somehow our `match()` check misses this and there's nothing I could do to mitigate that.. I understand now why you have used an additional group `suda_smart_read_local`.

I think I'll revert this to the way it was in your original suggestion, unless you have in mind a way to optimize my approach in a way that would actually work. Or, we can revert it to the way it was with the dictionary cache of buffer numbers checked which I still think is not that bad considering it is probably the most readable / understandable for anyone who may encounter this in the future..

Reply:
> But, the problem is, that since we have to use nested, and suda#BufEnter() calls edit, I always get double suda:// prefixes: suda://suda://%. Somehow our match() check misses this and there's nothing I could do to mitigate that.. I understand now why you have used an additional group suda_smart_read_local.

Probably `expand('<afile>')` is empty on `BufNew`. Try `BufAdd` instead.

Additionally, removing `suda_smart_read_secondary` is wrong while the second one should be applied to second, third, ... buffers as well (try `:e unreadable1` then `:e unreadable2` to see what I mean). So the correct one would be

```vim
if get(g:, 'suda_smart_read')
  augroup suda_smart_read_primary
    autocmd!
    autocmd BufEnter * nested call suda#BufEnter(expand('<afile>'))
    autocmd VimEnter *
          \ augroup suda_smart_read_primary |
          \   autocmd! * |
          \ augroup END
  augroup END

  augroup suda_smart_read_secondary
    autocmd!
    autocmd BufAdd * nested call suda#BufEnter(expand('<afile>'))
  augroup END
endif
```

But I'm not sure the following points

- Is `expand('<afile>')` returns value in `BufAdd` event?
    - May need `expand('<amatch>')`
- Is `file suda://% | edit` works as expect?

So above code might not work.

Reply:
That doesn't work just like my solution, unfortunately. I think this leaves us 2 options. Here they are summarized, along with the pros and cons IMO:

1. Use your original solution with both `augroup suda_smart_read_primary` and `augroup suda_smart_read_secondary` with `augroup suda_smart_read_local`.
   * Pros: Almost no unnecessary `autocmd`s (only `BufNew`).
   * Cons: Sort of complicated or in other words - very hard to understand for other developers when / if they'll read it.
2. Use the internal cache of buffers already checked.
   * Pros: Very simple to understand
   * Cons: ?

Anyway, Let me know what is your final opinion.

Reply:
OK. Try the followings. It should be better because `<C-o>/<c

```diff
diff --git a/autoload/suda.vim b/autoload/suda.vim
index 8487e31..069142f 100644
--- a/autoload/suda.vim
+++ b/autoload/suda.vim
@@ -94,32 +94,13 @@ function! suda#read(expr, ...) abort range
 endfunction
 
 function! suda#write_wrapper(write_cmd) abort
-  try 
+  try
     execute(a:write_cmd)
   catch /^Vim\%((\a\+)\)\=:E45:/
     execute(a:write_cmd . ' suda://%')
   endtry
 endfunction
 
-let s:smart_read_checked = {}
-
-function! suda#smart_read(bufn, fpath) abort
-  if get(s:smart_read_checked, a:bufn, 0)
-    let s:smart_read_checked[a:bufn] = 1
-    return
-  end
-  if !match(a:fpath, '^[a-z]\+://*') || &buftype != '' || a:fpath ==# ''
-    return
-  endif
-  if !filereadable(a:fpath)
-    if filewritable(fnamemodify(a:fpath, ':p:h')) ==# 2
-      return
-    endif
-    execute 'edit suda://%'
-    execute "bdelete! " . a:bufn
-  endif
-endfunction
-
 function! suda#write(expr, ...) abort range
   let path = s:expand_expression(a:expr)
   let options = extend({
@@ -161,7 +142,6 @@ function! suda#write(expr, ...) abort range
   endtry
 endfunction
 
-
 function! suda#BufReadCmd() abort
   call s:doautocmd('BufReadPre')
   let ul = &undolevels
@@ -226,6 +206,27 @@ function! suda#FileWriteCmd() abort
   endtry
 endfunction
 
+function! suda#BufEnter() abort
+  let bufname = expand('<amatch>')
+  if !empty(&buftype) || empty(bufname) || match(bufname, '^[a-z]\+://*') isnot# -1
+    " Non file buffer
+    return
+  elseif filereadable(bufname)
+    " File is readable
+    return
+  elseif empty(getftype(bufname)) && filewritable(fnamemodify(bufname, ':p:h')) is# 2
+    " File does not exist and the directory is writable
+    return
+  endif
+  let bufnr = bufnr('%')
+  execute printf(
+        \ 'keepalt keepjumps edit %s%s',
+        \ g:suda#prefix,
+        \ fnamemodify(bufname, ':p'),
+        \)
+  execute printf('silent! %dbwipeout', bufnr)
+endfunction
+
 function! s:escape_patterns(expr) abort
   return escape(a:expr, '^$~.*[]\')
 endfunction
diff --git a/plugin/suda.vim b/plugin/suda.vim
index 71aec8f..36d4f4c 100644
--- a/plugin/suda.vim
+++ b/plugin/suda.vim
@@ -15,8 +15,20 @@ if get(g:, 'suda_hijack_write')
 endif
 
 if get(g:, 'suda_smart_read')
-  augroup suda_smart_read
+  augroup suda_smart_read_primary
     autocmd!
-    autocmd BufEnter * nested call suda#smart_read(expand('<abuf>'), expand('<afile>'))
-  augroup end
+    autocmd BufEnter * nested call suda#BufEnter()
+    autocmd VimEnter *
+          \ augroup suda_smart_read_primary |
+          \   autocmd! * |
+          \ augroup END
+  augroup END
+
+  augroup suda_smart_read_secondary
+    autocmd! * <buffer>
+    autocmd BufNew *
+        \ augroup suda_smart_read_local |
+        \   autocmd BufEnter <buffer=abuf> nested call suda#BufEnter() |
+        \ augroup END
+  augroup END
 endif
```